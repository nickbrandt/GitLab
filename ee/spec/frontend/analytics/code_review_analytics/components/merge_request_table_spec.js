import { GlTable } from '@gitlab/ui';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import MergeRequestTable from 'ee/analytics/code_review_analytics/components/merge_request_table.vue';
import createState from 'ee/analytics/code_review_analytics/store/modules/merge_requests/state';
import { mockMergeRequests } from '../mock_data';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('MergeRequestTable component', () => {
  let wrapper;
  let vuexStore;

  const createStore = (initialState = {}) =>
    new Vuex.Store({
      modules: {
        mergeRequests: {
          namespaced: true,
          state: {
            ...createState(),
            ...initialState,
          },
        },
      },
    });

  const createComponent = (store) =>
    mount(MergeRequestTable, {
      localVue,
      store,
    });

  const bootstrap = (initialState) => {
    vuexStore = createStore(initialState);
    wrapper = createComponent(vuexStore);
  };

  afterEach(() => {
    wrapper.destroy();
  });

  const findTable = () => wrapper.findComponent(GlTable);
  const findTableRow = (index) => findTable().findAll('tbody tr').at(index);
  const findReviewTimeCol = (rowIndex) => findTableRow(rowIndex).findAll('td').at(1);

  const updateMergeRequests = (index, attrs) =>
    mockMergeRequests.map((item, idx) => {
      if (idx !== index) {
        return item;
      }

      return {
        ...item,
        ...attrs,
      };
    });

  describe('template', () => {
    beforeEach(() => {
      bootstrap({ mergeRequests: mockMergeRequests });
    });

    it('renders the GlTable component', () => {
      expect(findTable().exists()).toBe(true);
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders the correct table headers', () => {
      const tableHeaders = [
        'Merge Request',
        'Review time',
        'Author',
        'Approvers',
        'Comments',
        'Commits',
        'Line changes',
      ];
      const headers = findTable().findAll('th');

      expect(headers).toHaveLength(tableHeaders.length);

      tableHeaders.forEach((headerText, i) => expect(headers.at(i).text()).toEqual(headerText));
    });

    describe('review time column', () => {
      it('shows "days" when the review time is >= 24', () => {
        bootstrap({ mergeRequests: updateMergeRequests(0, { review_time: 64 }) });

        expect(findReviewTimeCol(0).text()).toBe('2 days');
      });

      it('shows "hours" when review time is < 24', () => {
        bootstrap({ mergeRequests: updateMergeRequests(0, { review_time: 18 }) });

        expect(findReviewTimeCol(0).text()).toBe('18 hours');
      });

      it('shows "< 1 hour" when review time is < 1', () => {
        bootstrap({ mergeRequests: updateMergeRequests(0, { review_time: 0 }) });

        expect(findReviewTimeCol(0).text()).toBe('< 1 hour');
      });

      it('shows "-" when review time is null', () => {
        bootstrap({ mergeRequests: updateMergeRequests(0, { review_time: null }) });

        expect(findReviewTimeCol(0).text()).toBe('â€“');
      });
    });
  });
});
