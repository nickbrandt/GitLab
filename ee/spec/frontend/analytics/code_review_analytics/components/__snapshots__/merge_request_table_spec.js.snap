// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeRequestTable component template matches the snapshot 1`] = `
<table
  aria-busy="false"
  aria-colcount="7"
  aria-describedby="__BVID__36__caption_"
  class="table b-table gl-table my-3 b-table-stacked-sm"
  id="__BVID__36"
  role="table"
>
  <!---->
  <!---->
  <thead
    class="thead-white border-bottom"
    role="rowgroup"
  >
    <!---->
    <tr
      class=""
      role="row"
    >
      <th
        aria-colindex="1"
        class="w-30p"
        role="columnheader"
        scope="col"
      >
        Merge request
      </th>
      <th
        aria-colindex="2"
        class="text-right"
        role="columnheader"
        scope="col"
      >
        <div>
          Review time
        </div>
      </th>
      <th
        aria-colindex="3"
        class=""
        role="columnheader"
        scope="col"
      >
        <div>
          Author
        </div>
      </th>
      <th
        aria-colindex="4"
        class=""
        role="columnheader"
        scope="col"
      >
        <div>
          Approvers
        </div>
      </th>
      <th
        aria-colindex="5"
        class="text-right"
        role="columnheader"
        scope="col"
      >
        <div>
          Comments
        </div>
      </th>
      <th
        aria-colindex="6"
        class="text-right"
        role="columnheader"
        scope="col"
      >
        <div>
          Commits
        </div>
      </th>
      <th
        aria-colindex="7"
        class="text-right"
        role="columnheader"
        scope="col"
      >
        <div>
          Line changes
        </div>
      </th>
    </tr>
  </thead>
  <tbody
    role="rowgroup"
  >
    <!---->
    <tr
      class=""
      role="row"
    >
      <td
        aria-colindex="1"
        class="table-col d-flex align-items-center"
        data-label="Merge request"
        role="cell"
      >
        <div>
          <div
            class="d-flex flex-column flex-grow align-items-end align-items-sm-start"
          >
            <div
              class="mr-title str-truncated my-2"
            >
              <a
                class="gl-link font-weight-bold text-plain"
                href="https://gitlab.com/gitlab-org/gitlab/merge_requests/38062"
                rel="noopener noreferrer"
                target="_blank"
              >
                This is just a super long merge request title that does not fit into one line so it needs to be truncated
              </a>
            </div>

            <ul
              class="horizontal-list list-items-separated text-secondary mb-0"
            >
              <li>
                !12345
              </li>

              <li>
                opened 5 months ago
              </li>

              <li>
                <span
                  class="d-flex align-items-center"
                >
                  <svg
                    aria-hidden="true"
                    class="mr-2 gl-icon s16"
                    data-testid="clock-icon"
                  >
                    <use
                      href="#clock"
                    />
                  </svg>

            11.1

                </span>
              </li>
            </ul>
          </div>
        </div>
      </td>
      <td
        aria-colindex="2"
        class="text-right table-col d-flex align-items-center d-sm-table-cell"
        data-label="Review time"
        role="cell"
      >
        <div>
          <span>
            2 days
          </span>
        </div>
      </td>
      <td
        aria-colindex="3"
        class="table-col d-flex align-items-center d-sm-table-cell"
        data-label="Author"
        role="cell"
      >
        <div>
          <a
            class="gl-link gl-avatar-link"
            href="https://gitlab.com/foo"
            rel="noopener noreferrer"
            target="_blank"
            title="foo"
          >
            <div
              class="gl-avatar gl-avatar-identicon gl-avatar-circle gl-avatar-s24 gl-avatar-identicon-bg1"
            >

  F

            </div>
          </a>
        </div>
      </td>
      <td
        aria-colindex="4"
        class="table-col d-flex align-items-center d-sm-table-cell"
        data-label="Approvers"
        role="cell"
      >
        <div>
          <div>
            <a
              class="gl-link gl-avatar-link"
              href="https://gitlab.com/bar"
              rel="noopener noreferrer"
              target="_blank"
              title="bar"
            >
              <div
                class="gl-avatar gl-avatar-identicon gl-avatar-circle gl-avatar-s24 gl-avatar-identicon-bg1"
              >



              </div>
            </a>
          </div>
        </div>
      </td>
      <td
        aria-colindex="5"
        class="text-right table-col d-flex align-items-center d-sm-table-cell"
        data-label="Comments"
        role="cell"
      >
        <div>
          21
        </div>
      </td>
      <td
        aria-colindex="6"
        class="text-right table-col d-flex align-items-center d-sm-table-cell"
        data-label="Commits"
        role="cell"
      >
        <div>
          <span>
            7
          </span>
        </div>
      </td>
      <td
        aria-colindex="7"
        class="text-right table-col d-flex align-items-center d-sm-table-cell"
        data-label="Line changes"
        role="cell"
      >
        <div>
          <span
            class="font-weight-bold cgreen"
          >
             +504
          </span>

          <span
            class="font-weight-bold cred"
          >
             -10
          </span>
        </div>
      </td>
    </tr>
    <!---->
    <!---->
  </tbody>
  <!---->
</table>
`;
